akka-persistence-typed/src/main/scala/akka/persistence/typed/Retention.scala

```diff
@@ -0,0 +1,55 @@
+package akka.persistence.typed
+
+/**
+  * Setup snapshot and event delete/retention behavior. Retention bridges snapshot
+  * and journal behavior. This defines the retention criteria.
+  *
+  * @param snapshotEveryNEvents Snapshots are used to reduce playback/recovery times.
+  *                             This defines when a new snapshot is persisted.
+  *
+  * @param keepNSnapshots      After a snapshot is successfully completed,
+  *                             - if 2: retain last maximum 2 *`snapshot-size` events
+  *                             and 3 snapshots (2 old + latest snapshot)
+  *                             - if 0: all events with equal or lower sequence number
+  *                             will not be retained.
+  *
+  * @param deleteEventsOnSnapshot Opt-in ability to delete older events on successful
+  *                               save of snapshot. Defaults to disabled.
+  */
+final case class RetentionCriteria(snapshotEveryNEvents: Long,
+                                   keepNSnapshots: Long,
+                                   deleteEventsOnSnapshot: Boolean) {
+  /**
+    * Delete Messages:
+    *   {{{ toSequenceNr - keepNSnapshots * snapshotEveryNEvents }}}
+    * Delete Snapshots:
+    *   {{{ (toSequenceNr - 1) - (keepNSnapshots * snapshotEveryNEvents) }}}
+    *
+    * @param lastSequenceNr the sequence number to delete to if `deleteEventsOnSnapshot` is false
+    */
+  def toSequenceNumber(lastSequenceNr: Long): Long = {
+    // Delete old events, retain the latest
+    lastSequenceNr - (keepNSnapshots * snapshotEveryNEvents)
+  }
+}
+
+object RetentionCriteria {
+
+  def apply(): RetentionCriteria =
+    RetentionCriteria(
+      snapshotEveryNEvents = 1000L,
+      keepNSnapshots = 2L,
+      deleteEventsOnSnapshot = false)
+
+  /** Scala API. */
+  def apply(snapshotEveryNEvents: Long, keepNSnapshots: Long): RetentionCriteria =
+    RetentionCriteria(snapshotEveryNEvents, keepNSnapshots, deleteEventsOnSnapshot = false)
+
+  /** Java API. */
+  def create(snapshotEveryNEvents: Long, keepNSnapshots: Long): RetentionCriteria =
+    apply(snapshotEveryNEvents, keepNSnapshots)
+
+  /** Java API. */
+  def create(snapshotEveryNEvents: Long, keepNSnapshots: Long, deleteMessagesOnSnapshot: Boolean): RetentionCriteria =
+    apply(snapshotEveryNEvents, keepNSnapshots, deleteMessagesOnSnapshot)
+}
```

Comment:
@patriknw I suggest we put any factory renames in the multiple algorithms PR, vs further work here. I can start that Monday, as an initial WIP for people to start considering - any design/impl changes.

Reply:
To clarify what I mean:

```
package akka.persistence.typed

import akka.annotation.InternalApi

object RetentionCriteria {

  /**
   * Snapshot and event delete/retention behavior. Retention bridges snapshot
   * and journal behavior. This defines the retention criteria.
   *
   * Snapshots are used to reduce playback/recovery times. By default it will retain the
   * last 2 snapshots and delete older snapshots. Events are by default not deleted.
   */
  val snapshotBasedRetentionCriteria: RetentionCriteria =
    SnapshotBasedRetentionCriteria(snapshotEveryNEvents = 1000L, keepNSnapshots = 2L, deleteEventsOnSnapshot = false)

  /**
   * Setup snapshot and event delete/retention behavior. Retention bridges snapshot
   * and journal behavior. This defines the retention criteria.
   *
   * @param snapshotEveryNEvents Snapshots are used to reduce playback/recovery times.
   *                             This defines when a new snapshot is persisted.
   *
   * @param keepNSnapshots      After a snapshot is successfully completed,
   *                             - if 2: retain last maximum 2 *`snapshot-size` events
   *                             and 3 snapshots (2 old + latest snapshot)
   *                             - if 0: all events with equal or lower sequence number
   *                             will not be retained.
   *
   * @param deleteEventsOnSnapshot Opt-in ability to delete older events on successful
   *                               save of snapshot. Defaults to disabled.
   */
  def snapshotBasedRetentionCriteria(
      snapshotEveryNEvents: Long,
      keepNSnapshots: Int,
      deleteEventsOnSnapshot: Boolean): RetentionCriteria =
    SnapshotBasedRetentionCriteria(snapshotEveryNEvents, keepNSnapshots, deleteEventsOnSnapshot)

}

/**
 * Snapshot and event delete/retention behavior.
 */
sealed trait RetentionCriteria

/**
 * INTERNAL API
 */
@InternalApi private[akka] final case class SnapshotBasedRetentionCriteria(
    snapshotEveryNEvents: Long,
    keepNSnapshots: Long,
    deleteEventsOnSnapshot: Boolean)
    extends RetentionCriteria {

  /**
   * Delete Messages:
   *   {{{ toSequenceNr - keepNSnapshots * snapshotEveryNEvents }}}
   * Delete Snapshots:
   *   {{{ (toSequenceNr - 1) - (keepNSnapshots * snapshotEveryNEvents) }}}
   *
   * @param lastSequenceNr the sequence number to delete to if `deleteEventsOnSnapshot` is false
   */
  def toSequenceNumber(lastSequenceNr: Long): Long = {
    // Delete old events, retain the latest
    lastSequenceNr - (keepNSnapshots * snapshotEveryNEvents)
  }

}
```

Then we can add more algorithms, such as
```
/**
 * INTERNAL API
 */
@InternalApi private[akka] final case class TimeBasedRetentionCriteria(
    deleteOlderThan: FiniteDuration,
    deleteEvents: Boolean)
    extends RetentionCriteria
```

Reply:
I create separate issue https://github.com/akka/akka/issues/26545 for this. Follow up in separate PR.

Reply:
Thanks

Reply:
It's similar in a way to what I have in my other branch from last week.